ingest/bin/post_process_metadata.py

```diff
@@ -0,0 +1,91 @@
+#! /usr/bin/env python3
+
+"""Reformat pandas DataTables for a pathogen build.
+
+Expecting one argument, a NCBI Virus metadata.tsv file
+"""
+# ===== Dependencies
+import argparse
+import os
+import sys
+
+import numpy as np
+import pandas as pd
+
+
+def parse_args():
+  # Main help command
+  parser = argparse.ArgumentParser(
+    description = "USAGE: python post_process_metadata.py --metadata 'data/metadata.tsv'"
+  )
+  # Add first argument
+  parser.add_argument(
+    "--metadata",
+    help = "NCBI Virus metadata.tsv file.",
+    required = True
+  )
+  parser.add_argument(
+    "--outfile", default="processed_metadata.tsv",
+    help = "Processed metdata file name. [default: processed_metadata.tsv] ",
+    required = False
+  )
+  
+  return parser.parse_args()
+
+# === Private methods
+
+def _set_strain_name(record):
+  """Check Isolate_s and Strain_s to find the strain name"""
+  strain_name=""
+  if record['strain'] != record['accession']:
+    strain_name=record['strain']
+  elif record['strain'] == record['accession'] and pd.notna(record['strain_s']):
+    strain_name=record['strain_s']
+  else:
+    strain_name=record['strain']
+
+  return strain_name.replace(' ', '_').replace('-',"_").replace(".","_").replace("(","_").replace(")","_")
+
+def _set_url(record):
+  """Set url column from accession"""
+  return "https://www.ncbi.nlm.nih.gov/nuccore/" + record['accession']
+
+def _set_paper_url(record):
+  """Set paper_url from publication"""
+  if pd.notna(record['publications']):
+    return "https://www.ncbi.nlm.nih.gov/pubmed/" + record['publications']
+  return ""
+
+def _set_dengue_serotype(record):
+  """Set dengue serotype from viruslineage_ids"""
+  if "11053" in record['viruslineage_ids']:
+    return "denv1"
+  elif "11060" in record['viruslineage_ids']:
+    return "denv2"
+  elif "11069" in record['viruslineage_ids']:
+    return "denv3"
+  elif "11070" in record['viruslineage_ids']:
+    return "denv4"
```

Comment:
I wonder if this mapping of lineage ids to serotype abbreviations could be represented by a top-level dictionary in the module instead of the conditional logic in this function. For example:

```python
serotype_by_lineage_id = {
    "11053": "denv1",
    "11060": "denv2",
    "11069": "denv3",
    "11070": "denv4",
}
```

The top-level dictionary exposes this information as configuration details that you can reuse later or quickly modify by changing the data structure instead of modifying the conditional logic code. This approach also opens the way to storing this configuration information in an external configuration file, for example, that you might prefer if you want to reuse this script for NCBI data for other viruses.

Without having looked at the NCBI metadata records before, this code also made we wonder what format the `viruslineage_ids` entries have. Are they usually a single id or do they store multiple ids with some standard delimiter? If the records only ever store a single id, you'll want to use a string test operator (`==`) instead of the `in` operator. For example, if you have a record with a lineage id of 110701, testing with `"11070" in "110701"` will evaluate to `True`, while `"11070" == "110701"` will not.

Even better, if your record only contains a single id and you've defined your mapping of serotype names by lineage ids as a dictionary, you can use a one-line call to [the pandas Series map function](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.map.html) instead of all of the conditional logic like so:

```python
def _set_dengue_serotype(record):
  """Set dengue serotype from viruslineage_ids"""
  return record.map(serotype_by_lineage_id)
```

Reply:
Yes! I'd prefer a generalizable dictionary as you've defined!

The`viruslineage_id` is a comma separated list of the NCBI Taxon IDs from general to more specific:

```
"10239,2559587,2732396,2732406,2732462,2732545,11050,11051,12637,11053"
```

To avoid the "11070" matching "110701", agree that the script should split `viruslineage_id` and do an exact match. Will look at record map, thanks!